File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys 
input = sys.stdin.readline 
sys.setrecursionlimit(10**6)
n, m = map(int, input().split())

parent = [0] * (n+1) 
for i in range(n+1): 
    parent[i] = i

def find(x):
    if parent[x] == x: 
        return x
    else:
        p = find(parent[x])
        parent[x] = p
        return p

def union(x, y):
    x, y = find(x), find(y)

    if x == y:
        return
    elif x<y:
        parent[y]=x 
    else:
        parent[x]=y
for _ in range(m):
    op, a, b = map(int, input().split())
    if op==0:
        union(a, b)
    elif op==1:
        if find(a)==find(b):
            print('YES')
        else:
            print('NO')

File: 백준/Gold/1717. 집합의 표현/집합의 표현.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
	parent = []int{}
	n      int
)

func find(x int) int {
	if parent[x] == x {
		return x
	}
	p := find(parent[x])
	parent[x] = p
	return p
}
func union(x, y int) {
	x, y = find(x), find(y)
	if x == y {
		return
	} else if x < y {
		parent[y] = x
	} else {
		parent[x] = y
	}
	return
}

func main() {
	defer writer.Flush()
	var m, op, a, b int
	fmt.Fscanf(reader, "%d %d\n", &n, &m)
	parent = make([]int, n+1)
	for i := 0; i < n+1; i++ {
		parent[i] = i
	}
	for i := 0; i < m; i++ {
		fmt.Fscanf(reader, "%d %d %d\n", &op, &a, &b)
		if op == 0 {
			union(a, b)
		} else if op == 1 {
			if find(a) == find(b) {
				fmt.Fprintf(writer, "YES\n")
			} else {
				fmt.Fprintf(writer, "NO\n")
			}
		}
	}
}
